data/transactions/logic/eval.go

```diff
@@ -4492,40 +4502,62 @@ func opAppGlobalDel(cx *EvalContext) error {
 // more than 2 or so, and was often called an "index".  But it was not a
 // basics.AssetIndex or basics.ApplicationIndex.
 
-func (cx *EvalContext) appReference(ref uint64, foreign bool) (basics.AppIndex, error) {
+func (cx *EvalContext) appReference(ref uint64, foreign bool) (aid basics.AppIndex, err error) {
 	if cx.version >= directRefEnabledVersion {
-		if ref == 0 || ref == uint64(cx.appID) {
-			return cx.appID, nil
-		}
-		aid := basics.AppIndex(ref)
-		if cx.availableApp(aid) {
-			return aid, nil
-		}
+		return cx.resolveApp(ref)
+	}
 
-		if cx.version < sharedResourcesVersion {
-			// Allow use of indexes, but this comes last so that clear advice can be
-			// given to anyone who cares about semantics in the first few rounds of
-			// a new network - don't use indexes for references, use the App ID
-			if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
-				return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
+	// resolveApp is already similarly protected (and must be, since it is
+	// called independently)
+	if cx.Proto.AppForbidLowResources {
+		defer func() {
+			if aid <= lastForbiddenResource && err == nil {
+				err = fmt.Errorf("low App lookup %d", aid)
 			}
+		}()
+	}
+	// Old rules
+	if ref == 0 { // Even back when expected to be a real ID, ref = 0 was current app
+		return cx.appID, nil
+	}
+	if foreign {
+		// In old versions, a foreign reference must be an index in ForeignApps or 0
```

Comment:
A lot of ambiguous usage of the word "old" in the comments

Reply:
Corrected to:
```
	// Old rules, pre directRefEnabledVersion, when a ref has to be a slot for
	// some opcodes, and had to be an ID for others.
```